### PR TITLE
feat: io preview via cookie

### PR DESCRIPTION
**Pull request by BRKalow**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-brkfeat-io-preview-cookie-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1201010428539925/1201722975655660/f) 🎟️

## What

Implements an adjusted version of @ashleemboyer's product switcher to drive switching between io sites for deploy previews.

## Why

This will allow us to preview io site changes easily!

## How

Implement rewrites using a cookie check. We look at the `io_preview` cookie's value and if it matches a product slug, we rewrite to its io site.

Note: this code will be thrown away at some point, and is largely a copy-paste of Ashlee's initial work, so I'm considering it to have a lower quality bar. 😅 

## Testing

- [ ] Navigate to the preview: https://dev-portal-git-brkfeat-io-preview-cookie-hashicorp.vercel.app/
- [ ] Select a product in the bottom left
- [ ] Validate that the page reloads and displays the product's io site

### CODE REVIEW

File: src/types/products.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+interface Product extends ProductMeta {
+  basePaths?: string[]
+  name: string
+  slug: ProductSlug
+  url?: string
+}
+
+type ProductGroup = Product[]
+
+export type { Product, ProductGroup, ProductSlug }
```
Comment (BRKalow) [inline src/types/products.ts:17]: largely copy pasted from the UI assembly branch.

File: src/components/_proxied-dot-io/common/preview-product-select/style.module.css [hunk 2]
```diff
@@ -0,0 +1,140 @@
+.productSwitcher {
```
Comment (BRKalow) [inline src/components/_proxied-dot-io/common/preview-product-select/style.module.css:1]: This file was largely copy pasted from the UI assembly branch.

File: src/components/_proxied-dot-io/common/preview-product-select/index.tsx [hunk 3]
```diff
@@ -0,0 +1,309 @@
+/**
```
Comment (BRKalow) [inline src/components/_proxied-dot-io/common/preview-product-select/index.tsx:1]: This file was largely copy pasted from the UI assembly branch. I adjusted the rendering to not be grouped and added some logic to set the cookie and reload on selection. It's also fixed in the corner now.

File: config/rewrites.js [hunk 4]
```diff
@@ -54,6 +54,18 @@ const dotIoRewrites = productsToProxy.reduce((acc, slug) => {
...
+    // To enable previewing of .io sites, we accept an io_preview cookie which must have a value matching a product slug
+    if (isPreview()) {
+      rewrite.has = [
+        {
+          type: 'cookie',
+          key: 'io_preview',
+          value: slug,
+        },
+      ]
+    }
```
Comment (BRKalow) [inline config/rewrites.js:67]: rewrites are ✨  magic

File: config/products.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import capitalize from '@hashicorp/platform-util/text/capitalize'
```
Comment (BRKalow) [inline config/products.ts:1]: This file is largely copy pasted from the UI assembly branch.

File: src/pages/_app.tsx [hunk 6]
```diff
@@ -14,6 +22,7 @@ export default function App({ Component, pageProps }) {
       <Layout>
         <Component {...pageProps} />
       </Layout>
+      {isPreview ? <PreviewProductSwitcher /> : null}
```
Comment (BRKalow) [inline src/pages/_app.tsx:25]: This ensures it only renders on preview deployments.

**Review by ashleemboyer — APPROVED** [review]
Works great other than the Sentinel option, so if that's expected you're good to go! 🥇

**Pull request merged**